### PR TITLE
BUG: fix DataFrame.apply returning wrong result when dealing with dtype (#28773)

### LINKED ISSUE #28773: apply sometimes unexpectantly casts int64 series to objects
#### Problem description

`pandas.DataFrame.apply()`seems to be converting series from int64 to object in some circumstances, and I'm not sure why. An example of the strange behavior I'm seeing is shown below, along with comments on what I'm expecting to see versus what I actually see. Note, this issue was originally reported on SO here: https://stackoverflow.com/questions/58222263/unexpected-behavior-when-applying-function-to-all-columns-in-pandas-data-frame.

#### Code Sample, a copy-pastable example if possible

```python
import pandas as pd
import numpy as np

df = pd.DataFrame({
    "col_1": [1, 2, 3],
    "col_2": ["hi", "there", "friend"]
})
print(df)
#>    col_1   col_2
#> 0      1      hi
#> 1      2   there
#> 2      3  friend
print(df.dtypes)
#> col_1     int64
#> col_2    object
#> dtype: object

# looks like np.issubdtype returns the expected result when calling the function
# on each series:
np.issubdtype(df.col_1, np.number)
#> True
np.issubdtype(df.col_2, np.number)
#> False

# but it doesn't return the expected result when using the apply function:
print(df.apply(lambda x: np.issubdtype(x, np.number)))
#> col_1    False
#> col_2    False
#> dtype: bool

# we can see that apply seems to be coercing the series to objects here:
print(df.apply(lambda x: x.dtype))
#> col_1    object
#> col_2    object
#> dtype: object

# what's also pretty weird is that i get the expected result when applying the
# replace_nulls() function below (e.g., median imputation is used if the series
# is a number, otherwise nulls are replaced with "MISSING"):
df = pd.DataFrame({
    "col_1": [1, 2, np.nan],
    "col_2": ["hi", "there", np.nan]
})

def replace_nulls(s):
    is_numeric = np.issubdtype(s, np.number)
    missing_value = s.median() if is_numeric else "MISSING"
    return np.where(s.isnull(), missing_value, s)

print(df.apply(replace_nulls))
#>    col_1    col_2
#> 0    1.0       hi
#> 1    2.0    there
#> 2    1.5  MISSING
```

<sup>Created on 2019-10-03 by the [reprexpy package](https://github.com/crew102/reprexpy)</sup>

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit           : None
python           : 3.6.5.final.0
python-bits      : 64
OS               : Darwin
OS-release       : 18.6.0
machine          : x86_64
processor        : i386
byteorder        : little
LC_ALL           : None
LANG             : None
LOCALE           : en_US.UTF-8
pandas           : 0.25.1
numpy            : 1.17.2
pytz             : 2019.2
dateutil         : 2.8.0
pip              : 19.0.3
setuptools       : 40.8.0
Cython           : None
pytest           : None
hypothesis       : None
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : None
html5lib         : None
pymysql          : None
psycopg2         : None
jinja2           : 2.10.1
IPython          : 7.8.0
pandas_datareader: None
bs4              : None
bottleneck       : None
fastparquet      : None
gcsfs            : None
lxml.etree       : None
matplotlib       : 3.1.1
numexpr          : None
odfpy            : None
openpyxl         : None
pandas_gbq       : None
pyarrow          : None
pytables         : None
s3fs             : None
scipy            : 1.3.1
sqlalchemy       : None
tables           : None
xarray           : None
xlrd             : None
xlwt             : None
xlsxwriter       : None

</details>

### PR DESCRIPTION
**Pull request by gorski-m**
- [x] closes #28773
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

The DataFrame.apply was sometimes returning wrong result when we passed
function, that was dealing with dtypes. It was caused by retrieving
the DataFrame.values of whole DataFrame, and applying the function
to it: values are represented by NumPy array, which has one
type for all data inside. It sometimes caused treating objects
in DataFrame as if they had one common type. What's worth mentioning,
the problem only existed, when we were applying function on columns.

The implemented solution "cuts" the DataFrame by columns and applies
function to each part, as it was whole DataFrame. After that, all
results are concatenated into final result on whole DataFrame.
The "cuts" are done in following way: the first column is taken, and
then we iterate through next columns and take them into first cut
while their dtype is identical as in the first column. The process
is then repeated for the rest of DataFrame.

### CODE REVIEW

**Comment by jbrockmendel** [comment]
>  The rest of changes are result of black pandas reformatting.

I guess you're using a different version of black than our CI is?  Regardless, can you revert non-relevant changes so we can focus on the important parts

**Comment by Reksbril** [comment]
@jbrockmendel done.

**Comment by jbrockmendel** [comment]
This is a lot of new code.  A couple things:

- Assuming this is the correct solution to this problem, the new code would belong in core.apply
- the "cuts" are a lot like the `DataFrame._data.blocks` that already exist.  Usually we discourage people from bothering with blocks but it is an option
- The hard part of the problem is identifying when you should be operating column-wise (i.e. not call .values).  Finding an elegant-ish way of doing that will be key here.

**Comment by Reksbril** [comment]
@jreback 
I've resolved a minor problem which caused most of tests to fail.

>     * the "cuts" are a lot like the `DataFrame._data.blocks` that already exist.  Usually we discourage people from bothering with blocks but it is an option
What's the benefit from using `DataFrame._data.blocks` instead of current solution? I'm just using DataFrame.iloc to get only part of DF and it seems to be very "clean" approach.
>     * The hard part of the problem is identifying when you should be operating column-wise (i.e. not call .values).  Finding an elegant-ish way of doing that will be key here.
I'm not sure if we can determine that at all. The obvious thing we can check is if all dtypes are the same (I think it's already done in current code and some specific apply variant is used in that case). As the functions given in input can be very complicated, the only solution to this problem I see, is to give some experimental input to the function, to determine if it relies on dtypes. However, that would be neither elegant nor easy to do (if even possible).

**Comment by jbrockmendel** [comment]
@Reksbril you're going to have to go through `frame_apply` and try to adapt the existing machinery rather than pile a new layer on top of it.  Within FrameApply there is a `apply_series_generator` that exists precisely to have a function operate column-by-column.  So a good solution to this problem would be to edit FrameApply.apply_standard so that this case goes through the apply_series_generator route.

File: pandas/core/frame.py [hunk 1]
```diff
@@ -6564,16 +6564,63 @@ def apply(self, func, axis=0, raw=False, result_type=None, args=(), **kwds):
...
-            self,
-            func=func,
-            axis=axis,
-            raw=raw,
-            result_type=result_type,
-            args=args,
-            kwds=kwds,
-        )
-        return op.get_result()
+        # Old apply function, which will be used for each part of DataFrame
```
Comment (jreback) [inline pandas/core/frame.py]: *all* of this should be done in pandas.core.apply

**Review by jreback — CHANGES_REQUESTED** [review]
pls delve deeper in how apply; we already have a column-by-column approach, your test case needs to inform when to use that

**Comment by Reksbril** [comment]
@jreback ping

**Comment by Reksbril** [comment]
@jreback ping

**Comment by jbrockmendel** [comment]
@Reksbril needs rebase, ill take a look

File: pandas/core/apply.py [hunk 2]
```diff
@@ -274,13 +274,16 @@ def apply_standard(self):
...
         # as demonstrated in gh-12244
-        if (
+        flag = (
             self.result_type in ["reduce", None]
             and not self.dtypes.apply(is_extension_array_dtype).any()
             # Disallow complex_internals since libreduction shortcut
             #  cannot handle MultiIndex
             and not isinstance(self.agg_axis, ABCMultiIndex)
-        ):
+        )
```
Comment (jbrockmendel) [inline pandas/core/apply.py]: what happens if you change this condition to also rule out mixed-dtypes?  that would prevent the call to self.values which is what does the object-dtype casting that is causing the problem

Comment (Reksbril) [reply pandas/core/apply.py]: Done. I forgot to add this additional rule before

File: pandas/core/apply.py [hunk 3]
```diff
@@ -274,13 +274,17 @@ def apply_standard(self):
...
         # as demonstrated in gh-12244
-        if (
+        flag = (
             self.result_type in ["reduce", None]
             and not self.dtypes.apply(is_extension_array_dtype).any()
             # Disallow complex_internals since libreduction shortcut
             #  cannot handle MultiIndex
             and not isinstance(self.agg_axis, ABCMultiIndex)
-        ):
+            and len(set(self.dtypes))
```
Comment (jreback) [inline pandas/core/apply.py]: use ```self.dtypes.nunique > 0```

call this something besides ``flag``, maybe ``can_reduce``

File: pandas/core/apply.py [hunk 4]
```diff
@@ -274,13 +274,17 @@ def apply_standard(self):
...
+        )
+        return_result = None
```
Comment (jreback) [inline pandas/core/apply.py]: why is this needed?

File: pandas/core/apply.py [hunk 5]
```diff
@@ -308,11 +312,19 @@ def apply_standard(self):
                 # reached via numexpr; fall back to python implementation
                 pass
             else:
-                return self.obj._constructor_sliced(result, index=labels)
+                return_result = self.obj._constructor_sliced(result, index=labels)
```
Comment (jreback) [inline pandas/core/apply.py]: what are you trying to do here? we have a specific subclass for Series where this could live, but I am not clear what you are trying to catch

File: pandas/core/apply.py [hunk 6]
```diff
@@ -308,11 +312,19 @@ def apply_standard(self):
...
+                if self.axis != 0 and self.axis != "index":
+                    return return_result
 
         # compute the result using the series generator
         results, res_index = self.apply_series_generator()
 
+        if flag and return_result is not None:
```
Comment (jreback) [inline pandas/core/apply.py]: what is this case tryng to catch?

File: pandas/tests/frame/test_apply.py [hunk 7]
```diff
@@ -691,6 +691,12 @@ def test_apply_dup_names_multi_agg(self):
 
         tm.assert_frame_equal(result, expected)
 
+    def test_apply_get_dtype(self):
+        # GH 28773
+        df = DataFrame({"col_1": [1, 2, 3], "col_2": ["hi", "there", "friend"]})
+        expected = Series(data=["int64", "object"], index=["col_1", "col_2"])
+        tm.assert_series_equal(df.apply(lambda x: x.dtype), expected)
```
Comment (jreback) [inline pandas/tests/frame/test_apply.py]: use 

result=
expected=
assert_.....

**Review by jreback — CHANGES_REQUESTED** [review]
your patch makes this much more complex; pls try to simplify

File: pandas/core/apply.py (hunk 6 above)
Comment (Reksbril) [reply pandas/core/apply.py]: There are 2 cases
1) can_reduce is False - the function doesn't change its behaviour, compared to previous version
2) can_reduce is True - the return_result is calculated like before, but it's returned only if we don't use apply on columns or we don't have mixed dtypes. In other case I use apply_series_generator to obtain column-by-column result and after that I'm wrapping it using _constructor_sliced. It's done this way to preserve the "old" way - it was solution to problem with my function sometimes returning Series instead of DataFrame (or the other way). I assumed, that already existing way of getting result would be better, than thinking about some new solutions.

File: pandas/core/apply.py (hunk 4 above)
Comment (Reksbril) [reply pandas/core/apply.py]: I'm using it to determine if line 314 was run.

**Comment by Reksbril** [comment]
@jreback ping

File: pandas/core/apply.py (hunk 6 above)
Comment (Reksbril) [reply pandas/core/apply.py]: @jreback

**Comment by WillAyd** [comment]
@Reksbril looks like a merge conflict and CI failure - can you fix those up and someone will take a look

File: pandas/core/apply.py [hunk 8]
```diff
@@ -274,13 +274,16 @@ def apply_standard(self):
...
             # Disallow complex_internals since libreduction shortcut
             #  cannot handle MultiIndex
             and not isinstance(self.agg_axis, ABCMultiIndex)
-        ):
+        )
+        return_result = None
+
+        if can_reduce:
 
             values = self.values
```
Comment (jbrockmendel) [inline pandas/core/apply.py]: @Reksbril the underlying problem here is that this call to `self.values` is casting to object dtype which you want to avoid.  if you add to the check in 277-283 `and self.obj._is_homogeneous_type` that should solve the whole thing

**Comment by jbrockmendel** [comment]
@Reksbril can you rebase and address comments

**Comment by Reksbril** [comment]
Yeah, I will. I'm sorry for long delay.

**Comment by Reksbril** [comment]
I won't be able to finish this for a few more months. The current version is the best I could do right now.

**Comment by simonjayhawkins** [comment]
> I won't be able to finish this for a few more months. The current version is the best I could do right now.

Thanks @Reksbril for working on this. closing to clear queue for now. ping when you want to continue and will reopen.
